### ISSUE TITLE
E1 blocker — pre-train evaluation never returns

### ISSUE THREAD

**Issue by celve** [haonan3/UniRLReport#1]
# E1 blocker — pre-train evaluation never returns

The run hangs forever in the baseline AIME evaluation, before rollout 0. No
training rollout has ever completed, so the 800-rollout budget is unspent.

**Not fixed.** Two earlier failures on the same launch were fixed; this one is
open.

---

## Signature — how to tell you have hit it

| | |
| --- | --- |
| last log line | `EVAL rollout 0 progress: batch=7 prompts=56 samples=896`, then nothing |
| GPUs | 32/32 at **0%**, memory still held |
| errors | **none** — no exception, no `ActorDiedError`, no OOM, exit code never returned |
| duration | indefinite; observed 18+ min before killing |

The process stays up and the log file stops growing. Any liveness check calls
this healthy.

---

## Environment

Reproduced twice on pod `unirl-4x8-gz-1-1788871208` (4 nodes × 8 H20 96 GB, gz),
image `mirrors.tencent.com/diffusionrl/diffusionrl:reproduce-multinode`.

| | |
| --- | --- |
| repo / commit | `celve/unirl` @ `4fbd3b4d` (branch `LIN-1171/e1-parity-gate`) |
| venv | `/root/unirl/.venv-sglang`, Python 3.12.13 |
| pins | torch 2.11.0+cu130 · sglang 0.5.12.post1 · ray 2.54.1 · transformers 5.6.0 |
| driver | 535.247.01, **glibc 2.28** |
| model | `/root/unirl/models/local/Qwen3-4B-Base` — pod-local **on every node** |
| train data | `data/dapo_math/train.jsonl` sha256 `fcc95077…f922c` (17,917 rows) |
| eval data | `data/dapo_math/aime_eval.jsonl` sha256 `4cf61fc8…a2ba` (**60 rows**) |

Three environment constraints, each of which independently breaks the run:

1. **Pin gcc-13 before `ray start`, not in the launcher.** A login shell resolves
   gcc 12, under which sglang's JIT activation kernel fails to compile
   (`Capture cuda graph failed: ninja exited with status 1`). Ray actors inherit
   the *daemon's* environment.
   ```bash
   export PATH="/opt/rh/gcc-toolset-13/root/usr/bin:$PATH"
   export CUDAHOSTCXX=/opt/rh/gcc-toolset-13/root/usr/bin/g++
   export NVCC_PREPEND_FLAGS="-ccbin /opt/rh/gcc-toolset-13/root/usr/bin/g++"
   ```
2. **CUDA 13 forward-compat layer** — the venv's `activate` prepends
   `/opt/cudacompat13/usr/local/cuda-13.3/compat`, so `cuDriverGetVersion`
   returns 13030 on driver 535. The sglang wheel is `manylinux_2_34` on a
   glibc-2.28 image, so it runs under `/opt/py312-glibc234`.
3. **Use `/root/unirl/.venv-sglang/bin/ray`,** not bare `ray` — conda's 2.46
   shadows the venv's 2.54 and the venv client refuses to join that cluster.
   Head on port **29500** (10000 is blocked between nodes), `--node-ip-address`
   from **bond1**.

---

## Reproduce

Ray cluster across all 4 nodes first (gcc pinned as above, in the daemon env):

```bash
RAY=/root/unirl/.venv-sglang/bin/ray
$RAY start --head --node-ip-address=<node0-bond1> --port=29500 --num-gpus=8   # node 0
$RAY start --address=<node0-bond1>:29500 --node-ip-address=<own-bond1> --num-gpus=8  # nodes 1-3
```

Then, from node 0, inside `tmux`:

```bash
export RAY_ADDRESS=<node0-bond1>:29500
export QWEN3_PATH=/root/unirl/models/local/Qwen3-4B-Base
export DATA_PATH=/root/unirl/data/dapo_math/train.jsonl
export EVAL_DATA_PATH=/root/unirl/data/dapo_math/aime_eval.jsonl
export UNIRL_MATHVERIFY_TIMEOUT_S=10

python -m unirl.train_ar --config-name=ar/qwen3_grpo_4b_base_dapo_sglang \
  num_devices=32 num_rollouts=800 eval_interval=10 \
  +save_interval=200 +save_dir=/dockerdata/e1_ckpt/<run> \
  ++data_source.args.run.seed=11 \
  logging.report_to_wandb=true logging.run_name=<run>
```

Hangs ~20 min in, during the `rollout_id = -1` baseline evaluation that
`eval_interval > 0` triggers **before** training. Setting `eval_interval=0`
skips the phase entirely — that is the untested discriminator, not a fix.

Note `save_interval`/`save_dir` need `+` (absent from the recipe); the seed field
is `data_source.args.run.seed` and needs `++`.

---

## Where it blocks

`unirl/trainer/ar.py:489`, inside `evaluate()`:

```python
486  generated = deep_hydrate(generated)
487  else:
488      generated = self.rollout.generate(sample)      # completed and returned
489  scored = self.reward.score_and_attach(generated)   # <-- blocks forever
```

`rollout_anchor_device` defaults to `None`, so `anchored` is False and the
`:488` branch runs. Driver stack:

```
get_objects (ray/_private/worker.py:984)
get (ray/_private/worker.py:2981)
_resolve_call (handle.py:674)
evaluate (ar.py:489)
train (ar.py:623)
```

A dispatched reward task whose result never returns. `ray.get` there takes no
timeout.

---

## Why batch 8, and why the recipe does not show it

The eval geometry comes from **`ARTrainer` defaults, not the recipe** — nothing
in `qwen3_grpo_4b_base_dapo_sglang.yaml` sets them (`ar.py:82-84`):

```
eval_num_prompts = -1     -> all 60 rows
eval_batch_size = 8       -> 8 batches; the last holds only 4 (60 = 7*8 + 4)
eval_samples_per_prompt = 16
```

So the final batch is a **4-prompt tail padded by `_pad_eval_inputs`**
(`ar.py:528`), and every eval batch fans out ×16. Both stalling attempts died
there. Reading the recipe alone will not tell you any of this.

---

## Diagnose

```bash
py-spy dump --pid $(pgrep -f "unirl.train_ar" | head -1)   # driver: expect ar.py:489
py-spy dump --pid $(pgrep -f "ray::Worker" | head -1)      # ranks: expect all idle
```

Confirm it is this and not attempt 1's cause — a **leaked grader thread** —
which presents identically at the GPU level but blocks at `:488`:

```bash
py-spy dump --pid <rank> | grep -c "Thread-.* (_run)"   # expect 0
py-spy dump --pid <rank> | grep -c "active+gil"         # expect 0
```

Nonzero means the grader is starving that rank's GIL — a different, fixed bug
(`4fbd3b4d`), not this one.

---

## Excluded, with evidence

| hypothesis | why excluded |
| --- | --- |
| SGLang generation / concurrency | driver blocks at `:489`; `:488` generate had already returned |
| Duplicate `sample_ids` on padded rows | `_pad_eval_inputs` mints a fresh root per pad row (`:eval-pad:i`) behind a collision guard, remapped via `map_sample_ids` |
| Reward-side sharding | padding is `lcm(rollout_dp, reward_dp)`, but `RewardService` has no `dp_size`, so the term is 1 — padding is rollout-driven |
| Grader GIL starvation | 0 leaked threads, 0 `active+gil` across all 18 ranks sampled |
| `_reset_pool()` holding `_POOL_LOCK` | structurally possible before `ddf57e3b`, now removed; pool children were idle, not wedged |

**Remaining:** a lost Ray task or dropped object reference on the padded tail.

---

## Contributing defect

Three call sites wait unbounded, which is why this presents as a silent hold
rather than an error:

- `unirl/trainer/ar.py:483` — eval generate (anchored path)
- `unirl/trainer/ar.py:489` — eval reward
- `unirl/rollout/engine/sglang/backends/native.py:103` — `future.result()`

---

## The three attempts

| run ID | commit | blocked at | cause | fixed |
| --- | --- | --- | --- | --- |
| `E1-unirl-qwen3-grpo-s11` | `e943d167` | `ar.py:488` | leaked grader thread spinning in sympy at 99.9% CPU holding the GIL | `4fbd3b4d` |
| `…-s11-r2` | `4fbd3b4d` | startup | `EADDRINUSE` on the sglang nccl port — stale sockets from a prior force-kill | teardown before relaunch |
| `…-s11-r3` | `4fbd3b4d` | `ar.py:489` | **unknown** | **no** |

`:488` and `:489` are adjacent but different calls; the first two rows are not
this bug.

---

## Why no probe caught it

Every pre-flight probe ran `eval_interval=0`. Twenty-five clean rollouts, three
parity runs and the full correctness gate never entered the evaluation path.

---

## Artifacts

`/mnt/gz/lin1171/` on the pod's shared Ceph: `formal/*/train.log` (all three
attempts), `RUN_MANIFEST.txt` (commit + overrides per run), `metrics_*.jsonl`,
`ray_logs/pyspy_stacks.txt` (stacks, transcribed verbatim — a live GIL-starved
rank cannot be recreated on demand).

Related: [stall investigation](E1-seed11-eval-stall.md) ·
[handover](E1-handover-to-E6.md) · [archive check](E1-archive-check.md)
